sensitivity-script.py

```diff
@@ -0,0 +1,72 @@
+import os
+import re
+import itertools
+import shutil
+
+# Path to the base config file
+base_config_file = "config.yml"
+backup_config_file = "config_backup.yml"
+
+# Backup the original config file to restore it later
+shutil.copyfile(base_config_file, backup_config_file)
+
+# Define the variables and their possible values
+variable_ranges = {
+    'fuel_cost_year': [2018,2019,2020,2021,2022,2023],
+    'load_growth': [0, 1],
+    'total_demand': [136e6, 185e6],
+    'atb_scenario': ['Moderate'] # 'Conservative', 'Advanced'
+}
+
+# Create a string of all the rules that must be forced to run
+forcerun = "retrieve_fuel_costs add_electricity" # add retrieve_costs when also modifying atb scenarios
+
+# Generate all possible combinations of the variables
+variable_combinations = list(itertools.product(
+    variable_ranges['fuel_cost_year'],
+    variable_ranges['load_growth'],
+    variable_ranges['total_demand'],
+    variable_ranges['atb_scenario']
+))
+
+# ATB scenarios must be handled differently because sub-parameters can't be passed to snakemake in the command line
+atb_scenario_targets = ["scenario: \'Moderate\'", "scenario: \'Conservative\'", "scenario: \'Advanced\'"]
+
+# Function to modify the atb_params in the config file (from ChatGPT)
+def replace_strings_in_config(file_path, target_strings, new_value):
+    with open(file_path, 'r', encoding='utf-8') as file:
+        content = file.read()
+
+    pattern = '|'.join(map(re.escape, target_strings))
+    updated_content = re.sub(pattern, new_value, content)
+
+    with open(file_path, 'w', encoding='utf-8') as file:
+        file.write(updated_content)
+
+run_no = 0
+for idx, (fuel_cost_year, load_growth, total_demand, atb_scenario) in enumerate(variable_combinations):
+    # Name the scenario
+    scenario = f"cost-{fuel_cost_year}_growth-{load_growth}_demand-{total_demand:.2E}_atb-{atb_scenario}"
+
+    # Combine all the top-line config variables
+    config = f"time_res=1 fuel_cost_year={fuel_cost_year} load_growth={load_growth} total_demand={total_demand} scenario={scenario}"
+
+    # Modify the config file to update the ATB scenario. 
+    # This is disabled because the model won't run -- alternate scenarios don't exist for nuclear
+    #atb_scenario_text = f"scenario: \'{atb_scenario}\'"
+    #replace_strings_in_config(base_config_file, atb_scenario_targets, atb_scenario_text)
+
+    # Run Snakemake, passing the new variables as configs
+    print(f"Running Snakemake with config {run_no} and the following command:")
+    print(f"snakemake --cores=4 --forcerun {forcerun} --config {config}")
+    os.system("snakemake --delete-all-output --cores=1")    # this could probably be optimized to not delete everything
```

Comment:
Is this step necessary? Seems like it will just delete the results from your run immediately after running the simulation. 

As long as you're not changing 
1. the time resolution
2. the number of model years

much of the timeseries data should be the same.

Reply:
So, the reason I added this is that snakemake keeps failing on various steps without it. It seems like snakemake isn't correctly determining when an input file has been changed, so it's not re-running rules that need to be re-run. Is there a better way to fix this? I can provide an example if needed.

Reply:
As long as you can keep all of the results from the model run then I don't mind that it's there. I'm just skeptical about it (but I haven't run it myself). 

If it works as you intended, then I'm okay with it.